Multisig: Connect request and SignMultisigTransaction request

sisou opened the pull request:
Related: https://github.com/nimiq/hub/pull/521

This PR adds two requests:
- Connect
- SignMultisigTransaction

## Connect

For this request, I am adding deterministic RSA key generation from the user's Key's `secret`. The public RSA key is then shared with the connected app, to be able to encrypt data that only the Keyguard can decrypt again. Encryption and decryption uses the `window.crypto.subtle` APIs.

## SignMultisigRequest

Nothing special added, this uses the code from the @nimiq/core SDK to do multisig signatures.

sisou committed:
Add test for signing a multisig transaction

sisou committed:
Add `signPartially` method to Key and use in test

sisou committed:
MonkeyPatch MerkleTree to verify multisig sender address

from participants' public keys

sisou committed:
[WIP] Add SignMultisigTransaction request

Missing multisig-specific UI elements

sisou committed:
Add demo for multisig signing

sisou committed:
Add multisig config badge to AddressInfo component

sisou committed:
Add user and account name section

sisou committed:
Demo of generating random RSA keys in a sandboxed iframe

sisou committed:
Use node-forge in sandboxed iframe to generate deterministic RSA key

sisou committed:
Use ArrayBuffers to exchange key material

sisou committed:
Extend 32-byte entropy to 1024-byte seed with PBKDF2

sisou committed:
Refactor Key config into an object and extend with rsaKeyPair option

sisou committed:
Add RSA key computation to Key class

sisou committed:
Implement encrypted secret parsing and decryption

sisou committed:
Add Rust code compiled to WASM for secret aggregation

sisou committed:
Update multisig badge UI, show identicon for legacy accounts

sisou committed:
Extract LoginFileAccountIcon into a component

sisou committed:
Refactor multisig request type structure

sisou committed:
Add comment about permission additions

sisou committed:
Narrow encryptionKey algorithm type

sisou committed:
Add sandboxed iframe files to dist during build

sisou committed:
Extract inline-script from RSAKeysIframe

Because CSP doesn't allow inline-scripts. This required getting the types for node-forge correct, too.

sisou committed:
Fix types for KeyguardCommand

sisou commented:
@danimoh Do you maybe have a better idea of how to solve the double `KeyguardCommand.KeyguardCommand` in this commit?
https://github.com/nimiq/keyguard/pull/458/commits/253219009034c528395025c8d42168deade52d9c

sisou committed:
Allow RSA key generation to use dynamic parameters

Requires passing the parameters in the request to ensure generation of the correct decryption key.

sisou committed:
Update title wording

sisou committed:
Add explainer tooltip to connect UI

sisou committed:
Connect request

sisou committed:
Clean-up, improve error handling

sisou committed:
Apply fixes from the Typescript update (PR #467)

sisou committed:
Allow signing multisig transactions from a vesting contract

sisou committed:
Update SignMultisigTransaction flow to Albatross PoS SDK

danimoh commented on .eslintignore:
```diff
@@ -1,3 +1,4 @@
 *.min.*
 src/lib/bitcoin/BitcoinJS.js
 src/lib/polygon/OpenGSN.js
+src/lib/multisig/wasm/pkg
```
This folder has been removed again in fbbcf067a5043d9d56bd24c5a2ac79f63355fac1.

danimoh commented on client/src/PublicRequest.ts:
```diff
@@ -189,6 +190,40 @@ export type SignTransactionRequest
+    senderLabel: string, // Not optional
+}> & {
+    multisigConfig: MultisigConfig,
+};
```
Could also be
```suggestion
export type SignMultisigTransactionRequestCommon = Transform<
    SignTransactionRequestCommon,
    'keyLabel' | 'senderLabel',
    Required<Pick<SignTransactionRequestCommon, 'keyLabel' | 'senderLabel'>>
> & {
    multisigConfig: MultisigConfig,
};
```
This ensures that there are no discrepancies if the original types of `keyLabel` or `senderLabel` were to change in `SignTransactionRequestCommon`, or if they were to be removed there. Obviously, in this case this benefit is of a rather theoretical nature. I just wanted to mention this alternative.

sisou replied:
Resolved in #524

danimoh commented on client/src/PublicRequest.ts:
```diff
@@ -94,6 +94,7 @@ export type BitcoinTransactionInfo = {
 };
 
 export type SignTransactionRequestLayout = 'standard' | 'checkout' | 'cashlink';
+export type SignMultisigTransactionRequestLayout = 'standard';
```
Are you planning to add other layouts?

sisou replied:
Who knows? This just keeps the option open.

danimoh commented on client/src/PublicRequest.ts:
```diff
@@ -593,8 +652,11 @@ export type Result = RedirectResult | IFrameResult;
 
 export type ResultType<T extends RedirectRequest> =
-    T extends Is<T, SignMessageRequest> | Is<T, SignTransactionRequest> ? SignatureResult :
+    T extends Is<T, SignMessageRequest> ? SignatureResult :
```
Should `SignMessageRequest` also receive an alias `SignMessageResult` for `SignatureResult`, similar to `SignMultisigTransactionResult`? Also in `RedirectResult` above and `ResultByCommand` below.

sisou replied:
No, instead I'll remove `SignMultisigTransactionResult` in #532 and type the result as `SignatureResult` instead to keep it simple.

danimoh commented on types/Keyguard.d.ts:
```diff
@@ -32,12 +32,38 @@ type AccountRecord = AccountInfo & {
+    kdf: string
+    iterations: number
+    keySize: number
+}
```
Can be used as `KeyguardRequest.EncryptionKeyParams` without copying it from `PublicRequest.ts`.

```suggestion
type EncryptionKeyParams = KeyguardRequest.EncryptionKeyParams;
```

sisou replied:
Resolved in #524

danimoh commented on package.json:
```diff
@@ -38,13 +38,14 @@
   "homepage": "https://github.com/nimiq/keyguard#readme",
   "devDependencies": {
-    "@nimiq/albatross-wasm": "npm:@nimiq/core@^2.0.6",
+    "@nimiq/albatross-wasm": "https://gitpkg.now.sh/nimiq/core-rs-albatross/web-client/dist?06999e1626b3312a906a0a490339d7feb781811b",
```
Will the Multisig primitives eventually be added to the regular `@nimiq/core` package?

sisou replied:
Yes, and this PR has been updated to use a regular released package version.

danimoh commented on tsconfig.json:
```diff
@@ -70,6 +70,7 @@
     "src/lib/QrScannerWorker.js",
     "src/lib/bitcoin/BitcoinJS.js",
     "src/lib/polygon/OpenGSN.js",
+    "src/lib/multisig/wasm/pkg",
```
This folder has been removed again in fbbcf067a5043d9d56bd24c5a2ac79f63355fac1.

danimoh commented on src/lib/rsa/sandboxed/forge.min.js.map:
Doesn't look like these mappings are very useful. (Even though they were taken directly from the `node-forge` package).

sisou replied:
Resolved in #524

danimoh commented on src/lib/rsa/sandboxed/forge.min.js:
For any other reviewer wondering: this is `forge.min.js` from `node-forge`  version 1.3.1 (sha256 dc67fd132427ad96c9666c844b39565413c40ddb1f2d063c53512fbf6d387dfd)

sisou replied:
I added a check for exactly this hash in the build script in #532

danimoh commented on src/lib/rsa/sandboxed/RSAKeysIframe.html:
```diff
@@ -0,0 +1,13 @@
+<html lang="en">
+<head>
+    <meta charset="UTF-8">
+    <title>RSA Keys Iframe | Keyguard Demo</title>
```
It's not only used in the demos.
```suggestion
    <title>RSA Keys Iframe</title>
```

sisou replied:
Resolved in #524

danimoh commented on src/lib/rsa/sandboxed/RSAKeysIframe.js:
```diff
@@ -0,0 +1,85 @@
+ * @returns {ArrayBuffer}
+ */
+function key2ArrayBuffer(key) {
+    const bytes = /** @type {string} */ (forge.asn1.toDer(key).getBytes());
```
Can be typed more accurately, and I'd personally prefer `keyToArrayBuffer` as a name (do we do that anywhere else to appreviate To as 2?). Additionally, our APIs mostly work on Uint8Arrays, so I'd suggest `keyToUint8Array`:
```js
/**
 * @param {forge.asn1.Asn1} key
 * @returns {Uint8Array}
 */
function keyToUint8Array(key) {
    const bytes = forge.asn1.toDer(key).getBytes();
    const result = new Uint8Array(bytes.length);
    for (let i = 0; i < bytes.length; i++) {
        result[i] = bytes.charCodeAt(i);
    }
    return result;
}
```

sisou replied:
Resolved in #524

danimoh commented on src/lib/rsa/sandboxed/RSAKeysIframe.html:
```diff
@@ -0,0 +1,13 @@
+    <meta charset="UTF-8">
+    <title>RSA Keys Iframe | Keyguard Demo</title>
+    <script src="./forge.min.js"></script>
+</head>
```
We should add a CSP meta tag here which enforces the sandbox, even when not opened in the iframe it was designed for, and set the `frame-ancestors` to only the keyguard to disallow the page being embedded in other pages.

Generally, the CSP for this iframe can be very strict, for example:
```html
    <meta http-equiv="Content-Security-Policy" content="
        sandbox allow-scripts;
        frame-ancestors 'self';
        default-src 'none';
        script-src 'self';
    ">
```

However, unfortunately, `sandbox` and `frame-ancestors` do not seem to be allowed to be specified via the `meta` tag. So those should be configured on the server by @curdbecker.

sisou replied:
Resolved in #524

danimoh commented on src/lib/rsa/sandboxed/RSAKeysIframe.js:
The entire script should refuse running (e.g. by throwing immediately) if it's not inside an iframe or if it's not an iframe within the Keyguard. Otherwise, a malicious page can open the page as iframe (unless disallowed by an appropriate CSP) or as a popup (always possible, I think), and communicate with it.

```js
if (// Refuse to run if:
    // - Run as top level window
    window.self === window.top
    // - Iframe is embedded in another domain
    || window.location.origin !== new URL(document.referrer).origin
    // - Iframe can access parent page and is thus not sandboxed
    || !!window.frameElement
) throw new Error('Only allowed to run in a sandboxed iframe');
```

sisou replied:
Resolved in #524

danimoh commented on src/lib/rsa/sandboxed/RSAKeysIframe.js:
```diff
@@ -0,0 +1,85 @@
+    if (event.source === event.target) {
+        // console.log("Ignored same-window event:", event);
+        return;
+    }
```
Where would such same-window events come from? Does `forge` send messages to the window itself?
(Mostly out of curiosity. Checking for it does no harm of course.)

sisou replied:
I think browser extensions could trigger these.

sisou replied:
I re-added the check in #532, after it was removed in #524.

danimoh commented on src/lib/rsa/sandboxed/RSAKeysIframe.js:
```diff
@@ -0,0 +1,85 @@
+    if (event.source === event.target) {
+        // console.log("Ignored same-window event:", event);
+        return;
+    }
```
Messages which are not coming from the Keyguard should be rejected.
```suggestion
    if (event.origin !== window.location.origin || !event.source) return; // Reject messages from other origins
```

sisou replied:
Resolved in #524

danimoh commented on src/lib/rsa/sandboxed/RSAKeysIframe.js:
```diff
@@ -0,0 +1,85 @@
+/* global forge */
+
+/**
+ * @typedef {{command: 'generateKey', seed: string, keySize: number}} GenerateKeyCommand
```
The `seed` should be better a `Uint8Array` instead of a binary string because it's not very clear that a binary string is expected, and not for example hex encoded data, and binary strings just feel very outdated.
```suggestion
 * @typedef {{command: 'generateKey', seed: Uint8Array, keySize: number}} GenerateKeyCommand
```

sisou replied:
Resolved in #524

danimoh commented on src/lib/rsa/sandboxed/RSAKeysIframe.js:
```diff
@@ -0,0 +1,85 @@
+        let seed = '';
+        while (seed.length < needed) {
+            seed += data.seed;
+        }
```
If not enough entropy was provided, we should throw and reject instead. The caller should make sure to provide enough entropy or stretch it out with a proper key derivation function, instead of us simply repeating the seed here.
```suggestion
        if (needed > seed.length) {
            throw new Error(`Seed does not provide enough entropy. ${needed} bytes needed`);
        }
```

sisou replied:
Resolved in #524

danimoh commented on src/lib/rsa/sandboxed/RSAKeysIframe.js:
```diff
@@ -0,0 +1,85 @@
+        while (seed.length < needed) {
+            seed += data.seed;
+        }
+        return seed.substring(0, needed);
```
There's no good reason to truncate the entropy if we have more than required, as whatever amount of entropy received [will be collected](https://github.com/digitalbazaar/forge/blob/a0a4a4264bedb3296974b9675349c9c190144aeb/lib/prng.js#L210).

Additionally, instead of truncating always to the first bytes of the entropy or returning the entire entropy, I was also thinking about returning a new slice of the entropy each time, because `forge.prng` does in fact [reseed for each generated random data](https://github.com/digitalbazaar/forge/blob/a0a4a4264bedb3296974b9675349c9c190144aeb/lib/prng.js#L90-L95). A quick experiment showed that a `forge.rsa.generateKeyPair` does in fact trigger two reseeds. Therefore one might consider returning two individual slices of `needed` bytes each, and overall requiring a seed of length `2 * needed`. However, looking at the linked comment, it seems like the frequent reseeding is mostly out of paranoia and not strictly required.

Overall, I would say returning the full, non-truncated seed in a single slice on all reseeds is a good middle ground.

sisou replied:
Resolved in #524

danimoh commented on src/lib/rsa/sandboxed/RSAKeysIframe.js:
```diff
@@ -0,0 +1,85 @@
+
+    //         resolve(keyPair);
+    //     },
+    // ));
```
> Sadly, those cannot be created in sandboxed iframes without `allow-same-origin`

Or luckily, because using webworkers could apparently cause the key generation to not be deterministic anymore: if a prng is passed, `generateKeyPair` uses a pure js implementation, which if async mode is enabled by passing a callback,  calls `_generateKeyPair`. `_generateKeyPair` in turn calls `prime.generateProbablePrime` which then calls `primeincFindPrime` and that calls `primeincFindPrimeWithWorkers` which features [the following comment](https://github.com/digitalbazaar/forge/blob/a0a4a4264bedb3296974b9675349c9c190144aeb/lib/prime.js#L150-L153):
```
// NOTE: This algorithm is indeterminate in nature because workers
// run in parallel looking at different segments of numbers. Even if this
// algorithm is run twice with the same input from a predictable RNG, it
// may produce different outputs.
```

Therefore, the async usage should be banned here.
```suggestion
    // Use the synchronous version of forge.rsa.generateKeyPair and not the asynchronous version with WebWorkers because
    // - WebWorkers cannot be created in sandboxed iframes without `allow-same-origin`:
    //   https://stackoverflow.com/a/30567964/4204380.
    // - Using multiple WebWorkers introduces race conditions, which causes the key generation to not be deterministic
    //   anymore, even if a deterministically seeded PRNG is used, see documentation on `primeincFindPrimeWithWorkers`:
    //   https://github.com/digitalbazaar/forge/blob/a0a4a4264bedb3296974b9675349c9c190144aeb/lib/prime.js#L150-L153
    //   If using WebWorkers, that method is called via `generateKeyPair` > `_generateKeyPair` > `generateProbablePrime`
    //   > `primeincFindPrime` > `primeincFindPrimeWithWorkers`.
    const keyPair = forge.pki.rsa.generateKeyPair({ bits: data.keySize, prng });
```

Unfortunately, this blocks the main thread in some browsers like Firefox, where same origin iframes are running on the same thread as the main window. In Chrome, [that is not the case](https://www.chromium.org/developers/design-documents/oop-iframes).

sisou replied:
Resolved in #524

danimoh committed:
Remove references to non-existent src/lib/multisig/wasm/pkg

That file was removed in fbbcf06.

danimoh committed:
PublicRequest,Keyguard.d.ts: small typing improvements

danimoh committed:
Remove forge.min.js.map as it contains no actually useful mappings

danimoh committed:
Key: security fix for RSA key seed generation

Previoulsy, the bytes have been interpreted as UTF8 bytes and encoded to a Javascript string,
which is UTF16 encoded. Up to 4 UTF8 bytes can represent a single unicode symbol, which is then
encoded as a single UTF16 character or a UTF16 surrogate pair. The following re-encodings are
possible:
1 UTF8 byte -> 1 UTF16 character
2 UTF8 bytes -> 1 UTF16 character
3 UTF8 bytes -> 1 UTF16 character
4 UTF8 bytes -> 2 UTF16 characters (surrogate pair)

This means that the resulting string could be as short as a third of the length of the input byte
array. As node-forge expects a binary string as input, where each character is interpreted as a
single byte, overall in the worst case, two thirds of the original data were discarded and an
exponentially weaker entropy was used.

This fix is a breaking change, as the resulting RSA key will be different. Therefore, users who
already have a stored RSA key need to logout to clear their old RSA key (or we decide to rename the
property in the store). Any app that already got the old RSA public key via a Connect request must
discard its known RSA public keys and associated encrypted data, and connect the account again to
get the new RSA public key.

danimoh commented on src/lib/rsa/sandboxed/RSAKeysIframe.js:
```diff
@@ -0,0 +1,85 @@
+    /** @type {Window} */ (event.source).postMessage({
+        privateKey: key2ArrayBuffer(privateKeyInfo),
+        publicKey: key2ArrayBuffer(publicKeyInfo),
+    }, '*');
```
```suggestion
    /** @type {Window} */ (event.source).postMessage({
        privateKey: key2ArrayBuffer(privateKeyInfo),
        publicKey: key2ArrayBuffer(publicKeyInfo),
    }, window.location.origin);
```
Either only allow sending the response to known keyguard origins, or the same origin the iframe is running on (also the default if no target origin is specified but the sandbox is messing with that so our `location.origin` has to be specified manually).

sisou replied:
Resolved in #524

danimoh committed:
Keys,RSAKeysIframe: use Uint8Buffers to communicate data

Binary strings are not very intuitive (and have been used incorrectly) and are not what
our APIs usually use.
Also now instead of copying the data, transfer ownership on postMessage communications.
Additionally, more thoroughly check the communicated data to be of the expected format.

danimoh commented on src/lib/Key.js:
```diff
@@ -11,16 +17,49 @@ class Key {
+     * @param {EncryptionKeyParams} paramsB
+     * @returns {boolean}
+     */
+    static _isEncryptionKeyParamsEqual(paramsA, paramsB) {
```
```suggestion
    /**
     * @private
     * @param {EncryptionKeyParams} paramsA
     * @param {EncryptionKeyParams} paramsB
     * @returns {boolean}
     */
    static _areEncryptionKeyParamsEqual(paramsA, paramsB) {
```

sisou replied:
Resolved in #524

danimoh commented on src/lib/Key.js:
```diff
@@ -11,16 +17,49 @@ class Key {
         this._id = null;
         this._defaultAddress = this.deriveAddress(Constants.DEFAULT_DERIVATION_PATH);
+
+        this.rsaKeyPair = config.rsaKeyPair;
```
```suggestion
        this._rsaKeyPair = config.rsaKeyPair;
```
For consistency with other properties, the name should start with an underscore, and then a getter be added. Also, I would move the definition just below the definition of `_hasPin`, as it also belongs to the `config`.

Additonal getter:
```js
    /**
     * @returns {RsaKeyPairExport | undefined}
     */
    getRsaKeyPairExportIfExists() {
        return this._rsaKeyPair;
    }
```

sisou replied:
Resolved in #524

danimoh commented on src/lib/Key.js:
```diff
@@ -50,6 +89,37 @@ class Key {
+        }, [
+            /** @type {Nimiq.PublicKey[]} */ ([]),
+            /** @type {Nimiq.Commitment[][]} */ ([]),
+        ]);
```
```suggestion
        const otherPublicKeys = otherSigners.map(({ publicKey }) => publicKey);
        const otherCommitments = otherSigners.map(({ commitments }) => commitments);
```

sisou replied:
Resolved in #524

danimoh commented on src/lib/Key.js:
```diff
@@ -11,16 +17,49 @@ class Key {
+     * @param {EncryptionKeyParams} params
+     * @returns {boolean}
+     */
+    static _isDefaultEncryptionKeyParams(params) {
```
```suggestion
    /**
     * @private
     * @param {EncryptionKeyParams} params
     * @returns {boolean}
     */
    static _areDefaultEncryptionKeyParams(params) {
```

sisou replied:
Resolved in #524

danimoh commented on src/lib/Key.js:
```diff
@@ -91,6 +161,124 @@ class Key {
+            if (Key._isDefaultEncryptionKeyParams(keyParams)) {
+                await KeyStore.instance.setRsaKeypair(this.id, this.rsaKeyPair);
+            }
+        }
```
This little snippet could be deduplicated into a method `setRsaKeyPair`.
```suggestion
        const rsaKeyPair = await this.setRsaKeyPair(keyParams);
```

With method:
```js
    /**
     * Calculates and sets the RSA key pair for the given key params. The RSA key is calculated deterministically based
     * on _secret, such that the same RSA key is generated for the same key params.
     * Warning: this method is computationally expensive.
     * @param {EncryptionKeyParams} keyParams
     * @returns {Promise<RsaKeyPairExport>}
     */
    async setRsaKeyPair(keyParams) {
        if (!this._rsaKeyPair || !Key._areEncryptionKeyParamsEqual(keyParams, this._rsaKeyPair.keyParams)) {
            this._rsaKeyPair = await this._computeRsaKeyPair(keyParams);
            if (Key._areDefaultEncryptionKeyParams(keyParams)) {
                await KeyStore.instance.setRsaKeypair(this.id, this._rsaKeyPair);
            }
        }
        return this._rsaKeyPair;
    }
```

sisou replied:
Resolved in #524

danimoh commented on src/lib/Key.js:
```diff
@@ -91,6 +161,124 @@ class Key {
+     * @returns {Promise<RsaKeyPairExport>}
+     */
+    async _computeRsaKeyPair(keyParams) {
+        const iframe = document.createElement('iframe');
```
Would be good to add a comment that explains why we're doing all this fun with the iframe, i.e. browser APIs don't support the deterministic generation of RSA keys from a seed, therefore we use a library, but to avoid an increased attack surface by including a library not audited by us, we run it in a sandbox.
```suggestion
        // Setup the RSAKeysIframe in which the actual computation of the RSA key happens via node-forge. The reason why
        // we don't use the browser crypto API instead, is that it does not support deterministic generation of RSA keys
        // from an existing secret, and the reason why node-forge is not used directly here is that we want to run it in
        // a sandbox, to reduce the attack surface from a big external dependency, not thoroughly audited by us.
        const iframe = document.createElement('iframe');
```

sisou replied:
Resolved in #524

danimoh committed:
RSAKeysIframe: add a content security policy

danimoh committed:
RSAKeysIframe: refuse running if not in a sandboxed iframe

Otherwise, a malicious page can open the page as iframe (unless disallowed by
an appropriate CSP) or as a popup (always possible, I think), and communicate
with it.

danimoh committed:
Key,RSAKeysIframe: security fix for postMessage communication with RSA iframe

Reject any messages in communication between the RSA iframe and main window, which are not
from / to the intended sender / recipient. Otherwise, the following attack was possible:
A malicious site starts a Connect request in an iframe (via Hub, which then redirects to the
Keyguard). Then, when the malicious site suspects that the Keyguard is waiting for the RSA key
calculation, it itself sends a postmessage to the popup with an RSA key known to the attacker,
which the Keyguard would willingly store and continue to use, also for other websites.

danimoh commented on src/lib/Key.js:
```diff
@@ -91,6 +161,124 @@ class Key {
+        iframe.setAttribute('sandbox', 'allow-scripts');
+        iframe.src = '../../lib/rsa/sandboxed/RSAKeysIframe.html'; // Relative path from a request URL
+        /** @type {Promise<void>} */
+        const loadPromise = new Promise(resolve => iframe.addEventListener('load', () => resolve()));
```
```suggestion
        const loadPromise = new Promise(resolve => iframe.addEventListener('load', resolve));
```

sisou replied:
Resolved in #524

danimoh committed:
RSAKeysIframe: refuse key generation if not enough entropy was provided

The caller should make sure to provide enough entropy or stretch it out with a proper
key derivation function, instead of us simply repeating the seed.

Additionally, don't truncate entropy if more was provided than necessary, as it is
been made use of nonetheless.

danimoh committed:
RSAKeysIframe: add notes on why RSA generation must be single threaded

danimoh commented on src/lib/Key.js:
```diff
@@ -91,6 +161,124 @@ class Key {
+        // Send computation command to iframe
+        iframe.contentWindow.postMessage({
+            command: 'generateKey',
+            seed: Nimiq.BufferUtils.toUtf8(seed), // seed is a bytestring
```
This is a very bad idea, and a security issue.

This interprets the bytes as UTF8 encoding and decodes it to a Javascript string, which is UTF16 encoded. Up to 4 UTF8 bytes can represent a single unicode symbol, which is then encoded as a single UTF16 character or a UTF16 surrogate pair. The following re-encodings are possible:
1 UTF8 byte -> 1 UTF16 character
2 UTF8 bytes -> 1 UTF16 character
3 UTF8 bytes -> 1 UTF16 character
4 UTF8 bytes -> 2 UTF16 characters (surrogate pair)

This means that the resulting string can be as short as a third of the length of the input byte array. As `node-forge` expects a binary string as input, where each character is interpreted as a single byte, overall in the worst case, two thirds of the original data are discarded and an exponentially weaker entropy is used.

Note that fixing this is a breaking change, as the resulting RSA key will be different. Therefore, users who already have a stored RSA key need to logout to clear their old RSA key (or we rename the property in the store). The multisig app must then discard its known RSA public keys and associated encrypted secrets, and connect the account again to get the new RSA public key.

sisou replied:
Resolved in #524

danimoh committed:
Key: some code cleanup

- Rename _isEncryptionKeyParamsEqual to _areEncryptionKeyParamsEqual and
  _isDefaultEncryptionKeyParams to _areDefaultEncryptionKeyParams and mark
  them as private.
- Make _rsaKey a private property. For external access add a method
  getRsaKeypairIfExists.
- Make collection of otherPublicKeys and otherCommitments in signPartially
  more concise.
- Deduplicate code for setting the RSA key into method setRsaKeyPair.
- Make usage of promises in _computeRsaKeyPair nicer.

danimoh committed:
Key: add documentation why we do RSA computations in a sandboxed iframe

danimoh committed:
Key: remove the RSA iframe after key generation

danimoh commented on src/lib/Key.js:
```diff
@@ -91,6 +161,124 @@ class Key {
+        // Listen for result from iframe
+        window.addEventListener('message', onMessage);
+
+        return resultPromise;
```
I think the Promise reads nicer like this:
```suggestion
        return new Promise(resolve => {
            /**
             * @param {MessageEvent} event
             */
            function onMessage(event) {
                if (event.source === event.target) {
                    // console.log("Ignored same-window event:", event);
                    return;
                }

                /** @type {{privateKey: ArrayBuffer, publicKey: ArrayBuffer}} */
                const data = event.data;
                if (!('privateKey' in data) || !('publicKey' in data)) return;

                window.removeEventListener('message', onMessage);

                resolve({
                    privateKey: new Uint8Array(data.privateKey),
                    publicKey: new Uint8Array(data.publicKey),
                    keyParams,
                });
            }

            // Listen for result from iframe
            window.addEventListener('message', onMessage);
        });
```

sisou replied:
Resolved in #524

danimoh commented on src/lib/Key.js:
```diff
@@ -91,6 +161,124 @@ class Key {
+            command: 'generateKey',
+            seed: Nimiq.BufferUtils.toUtf8(seed), // seed is a bytestring
+            keySize: keyParams.keySize,
+        }, '*');
```
Might want to add a comment that due to the iframe's sandboxed origin, we can't be more restrictive than `'*'`.

sisou replied:
Resolved in #524

danimoh commented on src/lib/Key.js:
```diff
@@ -91,6 +161,124 @@ class Key {
+            if (event.source === event.target) {
+                // console.log("Ignored same-window event:", event);
+                return;
+            }
```
Here, we absolutely must check that the message came from the iframe, otherwise the following attack is possible:
A malicious site starts a `Connect` request in an iframe (via Hub, which then redirects to the Keyguard). Then, when the malicious site suspects that the the Keyguard is waiting for the RSA key calculation, it itself sends a postmessage to the popup with an RSA key known to the attacker, which the Keyguard would willingly store and continue to use, also for other websites.

sisou replied:
Resolved in #524

danimoh commented on src/lib/Key.js:
```diff
@@ -91,6 +161,124 @@ class Key {
+            const data = event.data;
+            if (!('privateKey' in data) || !('publicKey' in data)) return;
+
+            window.removeEventListener('message', onMessage);
```
At this point, also the iframe should be removed from the DOM, such that it can be garbage collected.

sisou replied:
Resolved in #524

danimoh commented on src/lib/KeyStore.js:
```diff
@@ -147,12 +153,14 @@ class KeyStore {
             hasPin: key.hasPin,
             secret: buffer.subarray(0, buffer.byteLength),
             defaultAddress: key.defaultAddress.serialize(),
+            rsaKeyPair: key.rsaKeyPair,
```
I think if the key is password encrypted, we should encrypt the rsa key as well. Same in setRsaKeypair.

sisou replied:
Resolved in #524 & #526

danimoh commented on src/lib/RequestParser.js:
```diff
@@ -128,15 +128,16 @@ class RequestParser { // eslint-disable-line no-unused-vars
      */
     parseTransaction(object) {
+        if (!object || typeof object !== 'object' || object === null) {
+            throw new Errors.InvalidRequestError('Request must be an object');
```
```suggestion
        if (!object || typeof object !== 'object') {
            throw new Errors.InvalidRequestError('Transaction info must be an object');
```

- The `object === null` check is already covered by `!object`.
- The transaction is being parsed here, not the entire request.

sisou replied:
Resolved in #524

danimoh commented on tests/lib/RequestParser.spec.js:
```diff
@@ -189,12 +189,12 @@ describe('RequestParser', () => {
+        expect(() => requestParser.parseLogoUrl('file:///home/maxmustermann/image.jpg', true, 'shopLogoUrl')).toThrow();
 
-        expect(requestParser.parseShopLogoUrl(undefined)).toBe(undefined);
+        expect(requestParser.parseLogoUrl(undefined, true, 'shopLogoUrl')).toBe(undefined);
```
A test could be added here to test the `false` for the newly introduced `allowEmpty`.

sisou replied:
Added in #532

danimoh commented on tests/lib/Key.spec.js:
```diff
@@ -56,4 +56,58 @@ describe('Key', () => {
+    //     );
+
+    //     expect(partialSignatureB.toHex()).toEqual('caa6353261d250e2f1f67499f526c47503015e08d2a69169322fecae83cdf607');
+    // })
```
TODO

sisou replied:
Fixed and re-enabled in #532

danimoh commented on src/request/connect/index.html:
```diff
@@ -0,0 +1,133 @@
+                <span class="more">
+                    Connect and use your account to<br>
+                    approve transactions.
+                </span>
```
Missing i18n.

sisou replied:
Resolved in #524

danimoh committed:
I18n: support translation variables / placeholders

danimoh committed:
I18n: support translating to dom, including placeholders and auto lang updates

danimoh committed:
Obversable: fix handling of deleted listeners, e.g. for I18n observer

danimoh committed:
Connect: i18n fixes and improvements

- Add i18n support to some texts that were missing it.
- Translate full headline instead of individual fragments. This provides more
  context to translators and supports languages where the structure of the
  sentence needs to change.

danimoh committed:
Connect: some code cleanup

- Remove an unused button.
- Preserve stack trace on CoreError.
- Access key params without a necessary type assertion.
- Return meta data on the key algorithm directly from the key, instead of
  hardcoding it.
- Capitalization of error messages

danimoh commented on src/request/connect/index.html:
```diff
@@ -0,0 +1,133 @@
+
+                <div class="bottom-explainer">
+                    The connection is for approving only.<br>
+                    Your funds remain unaffected.
```
Missing i18n.

sisou replied:
Resolved in #524

danimoh commented on src/request/connect/index.html:
```diff
@@ -0,0 +1,133 @@
+
+            <div class="page-footer nq-card-footer">
+                <button data-i18n="passwordbox-connect-account" class="nq-button light-blue continue display-none">Connect your account</button>
+                <form id="password-box" class=""></form>
```
Empty `class` can be removed.

```suggestion
                <form id="password-box"></form>
```

sisou replied:
Resolved in #524

danimoh commented on src/request/connect/ConnectApi.js:
```diff
@@ -0,0 +1,75 @@
+    /** @type {KeyguardCommand.KeyguardCommand} */ ('derive-btc-xpub'),
+    // Swap
+    /** @type {KeyguardCommand.KeyguardCommand} */ ('sign-swap'),
+];
```
Should some of the new commands like `sign-staking` or `sign-polygon-transaction` be added?

```suggestion
const PermissionKeyguardCommands = [
    /** @type {KeyguardCommand.KeyguardCommand} */ ('remove-key'),
    /** @type {KeyguardCommand.KeyguardCommand} */ ('export'),
    /** @type {KeyguardCommand.KeyguardCommand} */ ('change-password'),
    // /** @type {KeyguardCommand.KeyguardCommand} */ ('sign-transaction'), // Already third-party whitelisted in Hub
    /** @type {KeyguardCommand.KeyguardCommand} */ ('sign-multisig-transaction'),
    // /** @type {KeyguardCommand.KeyguardCommand} */ ('sign-staking'), // Already third-party whitelisted in Hub
    // /** @type {KeyguardCommand.KeyguardCommand} */ ('sign-message'), // Already third-party whitelisted in Hub
    /** @type {KeyguardCommand.KeyguardCommand} */ ('derive-address'),
    // Bitcoin
    /** @type {KeyguardCommand.KeyguardCommand} */ ('sign-btc-transaction'),
    /** @type {KeyguardCommand.KeyguardCommand} */ ('derive-btc-xpub'),
    // Polygon
    /** @type {KeyguardCommand.KeyguardCommand} */ ('sign-polygon-transaction'),
    /** @type {KeyguardCommand.KeyguardCommand} */ ('derive-polygon-address'),
    // Swap
    /** @type {KeyguardCommand.KeyguardCommand} */ ('sign-swap'),
];
```

sisou replied:
Resolved in #524

danimoh committed:
Connect: provide more accurate information on exposed addresses

Previously, the UI wrongly claimed, that always the first address is exposed, which
is only the most common case. In fact though, an app can request arbitrary addresses
and an arbitrary number of addresses. In such case, the UI now shows the number of
requested addresses and their derivation paths.

danimoh commented on types/KeyguardCommand.d.ts:
```diff
@@ -0,0 +1,4 @@
+import * as _KeyguardCommand from '../client/src/KeyguardCommand';
+
+export as namespace KeyguardCommand;
+export = _KeyguardCommand;
```
Could be nice to actually reimplement the enum from the client in the Keyguard's js.

sisou replied:
Resolved in #524

danimoh commented on src/request/connect/index.html:
```diff
@@ -0,0 +1,133 @@
+            </div>
+
+            <div class="page-footer nq-card-footer">
+                <button data-i18n="passwordbox-connect-account" class="nq-button light-blue continue display-none">Connect your account</button>
```
Unused button

sisou replied:
Resolved in #524

danimoh commented on src/request/connect/Connect.js:
```diff
@@ -0,0 +1,139 @@
+        $button.addEventListener('click', () => {
+            $button.classList.add('display-none');
+            $passwordBox.classList.remove('display-none');
+        });
```
Unused event handler for unused buton.

sisou replied:
Resolved in #524

danimoh commented on src/request/connect/Connect.js:
```diff
@@ -0,0 +1,139 @@
+                this._passwordBox.onPasswordIncorrect();
+                return;
+            }
+            reject(new Errors.CoreError(errorMessage));
```
```suggestion
            reject(new Errors.CoreError(error instanceof Error ? error : errorMessage));
```
Passing the original error preserves the stack trace.

sisou replied:
Resolved in #524

danimoh commented on src/request/connect/Connect.js:
```diff
@@ -0,0 +1,139 @@
+
+            /** @type {Uint8Array} */
+            const messageBytes = Utf8Tools.stringToUtf8ByteArray(request.challenge);
+            const signature = key.signMessage(keyPath, messageBytes);
```
This blindly signs any arbitrary text message as a valid Nimiq message signature. I.e. an attacker can trick the user into unknowingly signing messages with the only restriction enforced by request parsing being that the message must be textual, which is the most common case for messages to be approved by the user. This can for example be abused to authenticate at staking pool websites as the user.

To avoid this, a prefix separate to the Nimiq message signature prefix should be used, for example `\x18Nimiq Signed Challenge:\n` (in contrast to `\x16Nimiq Signed Message:\n` used for Nimiq messages).
Alternatively, the challenge can still be signed as a Nimiq message, but with an additional prefix enforced on the challenge. A separate prefix is arguably more elegant.

This is a breaking change, which also requires adaptions in apps using the Connect request.

sisou replied:
Resolved in #524

danimoh commented on src/request/connect/index.html:
```diff
@@ -0,0 +1,133 @@
+                                The shared wallet has its own address. It can receive transactions from anywhere just like a regular wallet.
+                            </p>
+                            <p data-i18n="connect-tooltip-multisig-3">
+                                The first address of your account will be visible to all participants in your shared wallet.
```
Not the first address but the addresses for the `requestedKeyPaths`.

sisou replied:
Resolved in #524

danimoh commented on src/request/connect/ConnectApi.js:
```diff
@@ -0,0 +1,75 @@
+        parsedRequest.appName = this.parseAppName(request.appName);
+        parsedRequest.keyInfo = await this.parseKeyId(request.keyId);
+        parsedRequest.keyLabel = this.parseLabel(request.keyLabel);
+        parsedRequest.requestedKeyPaths = request.requestedKeyPaths.map(
+            path => this.parsePath(path, 'requestedKeyPaths'),
+        );
```
This can be abused by a malicious app to silently learn any number of the user's addresses. Those should probably be shown in the UI.

sisou replied:
Resolved in #524

danimoh commented on src/request/connect/Connect.js:
```diff
@@ -0,0 +1,139 @@
+        }
+
+        const rsaPublicCryptoKey = await key.getRsaPublicKey(Key.defaultEncryptionKeyParams);
+        const keyParams = /** @type {RsaKeyPairExport} */ (key.rsaKeyPair).keyParams;
```
```suggestion
        const keyParams = Key.defaultEncryptionKeyParams;
        const rsaPublicCryptoKey = await key.getRsaPublicKey(keyParams);
```

sisou replied:
Resolved in #524

danimoh commented on src/request/connect/Connect.js:
```diff
@@ -0,0 +1,139 @@
+            encryptionKey: {
+                format: 'spki',
+                keyData: new Uint8Array(await window.crypto.subtle.exportKey('spki', rsaPublicCryptoKey)),
+                algorithm: { name: 'RSA-OAEP', hash: 'SHA-256' },
```
Can read algorithm from `rsaPublicCryptoKey.algorithm` instead of hard coding the constant here.

```suggestion
                algorithm: {
                    name: rsaPublicCryptoKey.algorithm.name,
                    hash: (/** @type {RsaHashedKeyAlgorithm} */ (rsaPublicCryptoKey.algorithm)).hash.name,
                },
```

sisou replied:
Resolved in #524

danimoh committed:
Connect: security fix for blind signing of arbitrary messages

The Connect request blindly signed any arbitrary challenge as a valid Nimiq message
signature. I.e. an attacker could trick the user into unknowingly signing messages
with the only restriction enforced by request parsing being that the message must be
textual, which is the most common case for messages to be approved by the user. This
could be abused for impersonating the user and for example authenticating at staking
pool websites as the user.

To avoid this, a distinct prefix separate to the Nimiq message signature prefix is
now used.

It's a breaking change, which also requires adaptions to the signature verification
in apps using the Connect request.

This commit also improves the way message prefixes are shared between the Keyguard
client and the main Keyguard source code by redefining the message prefixes in a
type-checked fashion in the main code, instead of the client having to import it as
a javascript file and exposing the value in a hacky global variable on window.

danimoh committed:
Connect,KeyguardCommand: replicate KeyguardCommand as enum in js

Instead of working only with string values and a separate .d.ts file exporting
KeyguardCommand in an additional namespace, and only type asserting, but not type
checking the strings, replicate KeyguardCommand as a fully type-checked enum on
the js side.

danimoh committed:
ClientEnums: replicate client enums in a more concise fashion

Instead of type checking each enum value individually, typecheck the entire enum
definition at once, and define a utility function for enum replication for nicer
reading code.

danimoh committed:
Connect: add staking and polygon requests to requestable permissions

danimoh committed:
Connect: disable request permissions for now

The were only part of the API but not processed in a meaningful fashion:
They were not shown in the UI for confimation by the user, and were not persisted,
and requests did not check for granted permissions.

danimoh commented on client/src/PublicRequest.ts:
```diff
@@ -492,6 +527,24 @@ export type DerivePolygonAddressResult = {
 
+export type ConnectRequest = SimpleRequest & {
+    appLogoUrl: string,
+    permissions: KeyguardCommand[],
```
The permissions are not currently processed in a meaningful fashion. Neither are they displayed to the user, nor are they persisted or checked for. Therefore, this should be removed for now.

sisou replied:
They are processed, stored, and checked in the Hub.

danimoh replied:
But if the Hub is doing it, still this code does not need to be in Keyguard, because the Keyguard is not doing any of it.

danimoh commented on src/request/sign-multisig-transaction/index.html:
```diff
@@ -0,0 +1,154 @@
+    <link rel="stylesheet" href="../../components/LoginFileAccountIcon.css">
+    <!-- <link rel="stylesheet" href="../../components/Timer.css"> -->
+    <!-- <link rel="stylesheet" href="../../components/PaymentInfoLine.css"> -->
+    <link rel="stylesheet" href="SignMultisigTransaction.css">
```
`CurrencyInfo.js` and `FiatApi.js` are probably not necessary to be included.

sisou replied:
Resolved in #524

danimoh commented on src/request/sign-multisig-transaction/index.html:
```diff
@@ -0,0 +1,154 @@
+                        <span class="hide-as approving-with" data-i18n="sign-multisig-tx-approving-with">Approving with</span>
+                        <div class="login-file-account-icon"></div>
+                        <span class="account-name"></span>
+                    </div>
```
The structure of these sentences could be entirely different in other languages, and also the little text segments do not provide enough context to the translators of what the complete sentence will be in the end. Therefore, this should be complete texts with placeholders, which are then later replaced in Javascript and assembled as DOM elements.

sisou replied:
Resolved in #524

danimoh commented on src/request/sign-multisig-transaction/index.html:
```diff
@@ -0,0 +1,154 @@
+            <svg height="48" width="54" color="inherit" class="loading-spinner"><g>
+                <path class="big-hex" d="M51.9,21.9L41.3,3.6c-0.8-1.3-2.2-2.1-3.7-2.1H16.4c-1.5,0-2.9,0.8-3.7,2.1L2.1,21.9c-0.8,1.3-0.8,2.9,0,4.2 l10.6,18.3c0.8,1.3,2.2,2.1,3.7,2.1h21.3c1.5,0,2.9-0.8,3.7-2.1l10.6-18.3C52.7,24.8,52.7,23.2,51.9,21.9z" stroke="currentColor" stroke-width="3" fill="none" stroke-linecap="round" opacity="0.4" stroke-dasharray="92.5 60"/>
+                <path class="small-hex" d="M51.9,21.9L41.3,3.6c-0.8-1.3-2.2-2.1-3.7-2.1H16.4c-1.5,0-2.9,0.8-3.7,2.1L2.1,21.9c-0.8,1.3-0.8,2.9,0,4.2 l10.6,18.3c0.8,1.3,2.2,2.1,3.7,2.1h21.3c1.5,0,2.9-0.8,3.7-2.1l10.6-18.3C52.7,24.8,52.7,23.2,51.9,21.9z" stroke="currentColor" stroke-width="3" fill="none" stroke-linecap="round" stroke-dasharray="47.5 105"/>
+            </g></svg>
```
Could apply the same little optimization from 3b69993e here, too. Equally in `connect/index.html`.

sisou replied:
Resolved in #524

danimoh commented on src/request/sign-multisig-transaction/SignMultisigTransaction.css:
```diff
@@ -0,0 +1,210 @@
+    text-align: center;
+    max-width: 100%;
+    overflow-wrap: break-word;
+}
```
This entire section is almost identical to `SignTransaction.css`. Thus, that one should simply be imported instead of duplicating everything.

sisou replied:
Resolved in #524

danimoh commented on src/request/sign-multisig-transaction/SignMultisigTransaction.css:
```diff
@@ -0,0 +1,210 @@
+    font-size: 2rem;
+    gap: 0.5rem;
+    margin: 0 -2rem;
+    margin-bottom: -0.5rem;
```
```suggestion
    margin: 0 -2rem -0.5rem;
```

sisou replied:
Resolved in #524

danimoh commented on src/request/sign-multisig-transaction/SignMultisigTransactionApi.js:
```diff
@@ -0,0 +1,284 @@
+     * @returns {MultisigConfig}
+     */
+    parseMultisigConfig(object) {
+        if (!object || typeof object !== 'object' || object === null) {
```
```suggestion
        if (!object || typeof object !== 'object') {
```

sisou replied:
Resolved in #524

danimoh commented on src/request/sign-multisig-transaction/SignMultisigTransactionApi.js:
```diff
@@ -0,0 +1,284 @@
+                    return new Nimiq.RandomSecret(secret);
+                },
+            );
+        } else if ('encrypted' in object.secrets) {
```
```suggestion
        } else if (object.secrets && typeof object.secrets === 'object' && 'encrypted' in object.secrets && 'keyParams' in object.secrets) {
```

sisou replied:
Resolved in #524

danimoh commented on src/request/sign-multisig-transaction/SignMultisigTransactionApi.js:
```diff
@@ -0,0 +1,284 @@
+                    }
+                    return new Nimiq.RandomSecret(secret);
+                },
+            );
```
If `secret` cannot be deserialized as a valid `RandomSecret`, the error is not mapped to a`InvalidRequestError`.
```suggestion
            try {
                for (const secret of object.secrets) {
                    if (!(secret instanceof Uint8Array)) throw new Error('must be Uint8Arrays');
                    secrets.push(new Nimiq.RandomSecret(secret));
                }
            } catch (error) {
                const errorMessage = error instanceof Error ? error.message : String(error);
                throw new Errors.InvalidRequestError(`Invalid secrets: ${errorMessage}`);
            }
```

sisou replied:
Resolved in #524

danimoh commented on src/request/sign-multisig-transaction/SignMultisigTransactionApi.js:
```diff
@@ -0,0 +1,284 @@
+                throw new Errors.InvalidRequestError(
+                    'Invalid secrets.encrypted: must be an array with at least 2 elements',
+                );
+            }
```
The length check is somewhat unnecessary and also not done above for unencrypted secrets.
Instead, a more sound check would be for `SignTransaction` to check that the number of my commitments, which it identifies after deriving my public key, matches the number of my secrets, which it currently doesn't do. If it would do that, no further check would be required that it should be at least two, because the commitments were all already checked to be at least two.

sisou replied:
But the check against the number of commitments only happens after password entering, whereas this check happens at the start of the request. The earlier the better.

danimoh replied:
There is not really an advantage in an early check. The check is to protect from invalid or malicious requests. In production, a working app should not be issuing invalid requests. During development, the dev might encounter the error during testing, and in that case indeed only after password entry, but for the dev that's fine. He is supposed to fix the issue after encountering it once. As for malicious requests, as the request is still eventually rejected, there is nothing to be gained, such that a malicious request for this error is also not to be expected in production.
I.e. for production, the additional, early check just bloats the code.

danimoh commented on src/request/sign-multisig-transaction/SignMultisigTransactionApi.js:
```diff
@@ -0,0 +1,284 @@
+            }
+            // Validate keyParams
+            if (!object.secrets.keyParams) {
+                throw new Errors.InvalidRequestError('Missing secrets.keyParams');
```
```suggestion
            if (!object.secrets.keyParams || typeof object.secrets.keyParams !== 'object') {
                throw new Errors.InvalidRequestError('Invalid secrets.keyParams');
```

sisou replied:
Resolved in #524

danimoh committed:
SignMultisigTransactionApi: stricter and improved request parsing

- Remove redundant null checks.
- Use existing type MultisigConfig['signers'] instead of redefining it.
- For signers, change an InvalidRequestError to a regular Error, as it will be
  mapped to an InvalidRequestError on catch anyways.
- For secrets, ensure that any thrown error is mapped to an InvalidRequestError,
  especially from the RandomSecret constructor.
- More thoroughly check structure of secrets object for encrypted secrets: check
  that it's an object, non null, has the expected properties, and that keyParams
  is an object.
- Use Array.every instead of Array.some to not have to negate the condition.
- Make multisig address validation more concise and more readable with added
  comments.
- Avoid a @ts-ignore in layout parsing.

danimoh commented on src/request/sign-multisig-transaction/SignMultisigTransactionApi.js:
```diff
@@ -0,0 +1,284 @@
+     * @param {MultisigConfig} multisig
+     */
+    verifyMultisigAddress(transaction, multisig) {
+        const multisigAddress = Nimiq.Address.fromPublicKeys(
+            multisig.publicKeys,
+            multisig.signers.length,
+        );
```
This limits the Multisig support to k of n Multisigs, where each combination of signers which can sign together has identical size k and draws from the n public keys. More advanced combinations are not supported, for example of varying size, e.g. a Multisig which can be signed by single master key A, or keys B and C combined, or combinations with no overlap. But that's fine. If ever needed, support for such can be added in the future by instead of passing the `publicKeys`, passing something like `publicKeyCombinations`.
Maybe some documentation should be added to `PublicRequestTypes` explaining what the publicKeys represent (i.e. the n public keys of a k of n multisig, and not aggregated public keys calculated over pre-determined combinations), and that only k of n Multisigs are supported currently.

sisou replied:
Resolved in #524

danimoh commented on src/request/sign-multisig-transaction/SignMultisigTransactionApi.js:
```diff
@@ -0,0 +1,284 @@
+            throw new Errors.InvalidRequestError(
+                'The multisig account must either be the sender or the recipient of the transaction',
+            );
+        }
```
More concise and with documentation:
```suggestion
        if (!(transaction.senderType === Nimiq.AccountType.Basic && transaction.sender.equals(multisigAddress))
            // Vesting contracts owned by the multisig are allowed to send anywhere. The contract address can't be
            // verified here, it can however be assumed to be owned by the multisig, as only the owner can sign a valid
            // vesting withdrawal transaction.
            && transaction.senderType !== Nimiq.AccountType.Vesting
            // For other transactions which are not from a basic account, i.e. from a contract (other than a vesting
            // contract), we enforce the multisig to be the recipient. E.g. unstaking, HTLC refund, HTLC withdrawal etc.
            // are only allowed to the multisig address.
            && !(transaction.recipientType === Nimiq.AccountType.Basic && transaction.recipient.equals(multisigAddress))
        ) {
            throw new Errors.InvalidRequestError(
                'The transaction must either be send from or to the multisig, or from a vesting contract owned by it',
            );
        }
```

sisou replied:
I do not agree with the more concise version, its not understandable. The documentation can also be added to the more verbose layout.

danimoh replied:
I think the more concise version is actually better to understand, which is why I changed it. It clearly lists the acceptable cases, and does not mix acceptable and unacceptable cases.

danimoh committed:
SignMultisigTransaction: remove unused dependencies CurrencyInfo and FiatApi

danimoh committed:
SignMultisigTransaction: some code cleanup

- Remove an unnecessary eslint-disable rule.
- Remove a duplicate newline.
- Remove unnecessary clearing of login file icon's HTML content.
- Preserve stack trace on CoreError.
- Deduplicate loop for collecting own commitment pairs.
- Read key algorithm from key, instead of hardcoding it.
- Simplify filtering other signers.

danimoh commented on src/request/sign-multisig-transaction/SignMultisigTransaction.js:
```diff
@@ -0,0 +1,309 @@
+            new LoginFileAccountIcon(request.keyInfo.defaultAddress.toUserFriendlyAddress(), $loginFileIcon);
+        } else {
+            // Show identicon for legacy accounts (which must be supported to support Team Nimiq Multisig)
+            $loginFileIcon.innerHTML = ''; // Remove LoginFile icon
```
The HTML element is already empty.

sisou replied:
Resolved in #524

danimoh committed:
SignMultisigTransaction: i18n improvements

Translate name section as full sentence instead of individual fragments.
This provides more context to translators and supports languages where the
structure of the sentence needs to change.

danimoh committed:
Connect,SignMultisigTransaction: reduce repetition in loading spinner SVG

Apply the same little optimization from 3b69993 to the Connect and
SignMultisigTransaction requests.

danimoh commented on src/request/sign-multisig-transaction/SignMultisigTransaction.js:
```diff
@@ -0,0 +1,309 @@
+                    ownSigner.commitments[i],
+                ));
+            }
+        }
```
The `for` loop can be deduplicated by first collecting only the `secrets` in the `if ... else ...` block, and doing the loop afterwards.
```suggestion
        /** @type {Nimiq.RandomSecret[]} */
        let ownCommitmentSecrets;
        if (Array.isArray(request.multisigConfig.secrets)) {
            ownCommitmentSecrets = request.multisigConfig.secrets;
        } else {
            // If we only have encrypted secrets, decrypt them first
            const rsaKey = await key.getRsaPrivateKey(request.multisigConfig.secrets.keyParams);

            try {
                ownCommitmentSecrets = await Promise.all(request.multisigConfig.secrets.encrypted.map(
                    async encrypted => new Nimiq.RandomSecret(new Uint8Array(
                        await window.crypto.subtle.decrypt(rsaKey.algorithm, rsaKey, encrypted),
                    )),
                ));
            } catch (error) {
                const errorMessage = error instanceof Error ? error.message : String(error);
                reject(new Errors.InvalidRequestError(`Cannot decrypt secrets: ${errorMessage}`));
                return;
            }
        }
        /** @type {Nimiq.CommitmentPair[]} */
        const ownCommitmentPairs = [];
        for (let i = 0; i < ownCommitmentSecrets.length; i++) {
            ownCommitmentPairs.push(new Nimiq.CommitmentPair(
                ownCommitmentSecrets[i],
                ownSigner.commitments[i],
            ));
        }
```

sisou replied:
Resolved in #524

danimoh commented on src/request/sign-multisig-transaction/SignMultisigTransaction.js:
```diff
@@ -0,0 +1,309 @@
+                signers: request.multisigConfig.signers.length,
+                participants: request.multisigConfig.publicKeys.length,
+            },
+        });
```
If the multisig is the recipient instead of the sender, the `multisig` property should be on the `_recipientAddressInfo`.

sisou replied:
Resolved in #524

danimoh commented on src/request/sign-multisig-transaction/SignMultisigTransaction.js:
```diff
@@ -0,0 +1,309 @@
+
+        if (transaction.flags === Nimiq.TransactionFlag.ContractCreation) {
+            // TODO: Decode contract creation transactions
+            // return ...
```
Even if not fully decoding the transaction data, it should at least hint at this being an interaction with a contract, be it contract creation, withdrawal, staking etc, based on sender and recipient type. Also even if there is no recipient data present.

sisou replied:
Resolved in #524

danimoh commented on src/components/AddressInfo.js:
```diff
@@ -58,6 +69,18 @@ class AddressInfo { // eslint-disable-line no-unused-vars
+            $badge.classList.add('multisig-badge', 'nq-blue-bg');
+            $badge.innerHTML = TemplateTags.hasVars(2)`
+                ${this._addressInfo.multisig.signers}
+                <span class="of">of</span>
```
Missing i18n.

sisou replied:
Resolved in #524

danimoh commented on src/components/AddressInfo.css:
```diff
@@ -37,6 +38,28 @@
+    padding: 0.375rem 0.75rem;
+    line-height: 1;
+    border: 0.5rem solid white;
+    word-spacing: -0.071428571em; // 1/14
```
Invalid CSS comment (should be `/* 1/14 */`). But probably better as `calc(-1em / 14)` anyways.
```suggestion
    word-spacing: calc(-1em / 14);
```

sisou replied:
Resolved in #524

danimoh commented on src/components/LoginFileAccountIcon.js:
```diff
@@ -0,0 +1,57 @@
+
+class LoginFileAccountIcon { // eslint-disable-line no-unused-vars
+    /**
+     * @param {string} [address]
```
Why is the address allowed to be `undefined`? It's nowhere used in such a fashion.

sisou replied:
Resolved in #524

danimoh commented on tools/build.sh:
```diff
@@ -365,6 +366,7 @@ cp -v favicon.ico dist
 cp -rv src/assets/* dist/assets/
 cp -v src/lib/QrScannerWorker.js dist/lib/
 cp -v node_modules/@nimiq/style/nimiq-style.icons.svg dist/assets/
+cp -v src/lib/rsa/sandboxed/* dist/lib/rsa/sandboxed/
```
Integrity hashes on the js files would be good.

sisou replied:
Resolved in #524

danimoh commented on demos/index.html:
```diff
@@ -24,6 +24,9 @@
 <div><a href="/demos/RemoveKey.html">Remove</a></div>
 <div><a href="/demos/SignTransaction.html">Sign TX</a></div>
 <div><a href="/demos/DerivePolygonAddress.html">Derive Polygon Address</a></div>
+<div><a href="/demos/SignMultisigTransaction.html">Sign Multisig TX</a></div>
+<div><a href="/demos/RSAKeys.html">RSA Keys</a></div>
+<div><a href="/demos/ConnectAccount.html">Connect Account</a></div>
```
These demos don't work currently.

sisou replied:
Fixed in #532

danimoh committed:
SignMultisigTransaction: deduplicate style between SignTransaction and SignMultisigTransaction

danimoh committed:
SignMultisigTransaction: show multisig badge on correct side of the transaction

If the multisig is the recipient of the transaction, for example for a remove-stake
transaction, show the multisig badge on the recipient side.

danimoh committed:
SignTransaction,SignMultiSigTransaction: support decoding advanced tx data

- Don't hide information on sender type, recipient type or sender data from the user.
  Instead, try to represent them in a user-friendly fashion.
- Decode transaction data like contract creation data or staking data, instead of
  displaying them only as hex.
- Allow decoding data as utf8 only for regular transactions.
- Provide info on contract withdrawals.
- Deduplicate code for transaction data formatting.
- Update formatted data on language change.

danimoh committed:
SignTransactionApi: for staking tx disallow the SignTransaction request

The SignStaking request should be used instead, which provides a specialized UI for it.

danimoh committed:
AddressInfo: i18n for multisig badge

danimoh committed:
AddressInfo.css: remove invalid css comment and hardcoded value

danimoh committed:
LoginFileIcon: make the address mandatory in the constructor

The class was not used anywhere without providing the address in the constructor.

danimoh committed:
build.sh: add integrity checks for RSAKeysIframe

danimoh committed:
PublicRequest: add documentation on MultisigConfig

danimoh committed:
PublicRequest,Keyguard.d.ts,Keys: rename EncryptionKeyParams to RsaKeyParams

- Avoid confusion with other encryption algorithms used in KeyStore.
- RSA is the only encryption algorithm supported by Connect and SignMultisigTransaction.
  Although the APIs / types are partially designed for supporting different algorithms, they
  aren't really, not entirely and not consequently designed for it, and the name change
  also doesn't change anything about the possibility to support additional algorithms in
  the future.
- More consistent naming with other RSA related methods.
- Rename _areEncryptionKeyParamsEqual to _areEqualRsaKeyParams and _areDefaultEncryptionKeyParams
  to _areDefaultRsaKeyParams.

danimoh committed:
KeyStore: store RSA key encrypted

danimoh committed:
SignMultisigTransaction: check that number of commitments and secrets match

danimoh committed:
KeyStore,Key: use AES for RSA key encryption

danimoh reviewed (changes requested):
In my opinion, quite a bit too many shortcuts have been taken, for example regarding missing i18n, unencrypted storage of private keys and omission of of relevant information in the UI.

There were also several serious security issues:
- generated RSA keys are orders of magnitude weaker than they should be, due to the way they are generated.
- (almost) arbitrary Nimiq messages are blindly signed, which can be used to impersonate the user.
- malicious apps are able to trick the Keyguard into storing and using RSA private keys known to them.

Additionally, this PR points out opportunities for further security hardening, code cleanup, code deduplication, smaller fixes, stricter parsing etc..

danimoh reviewed (changes requested):
In my opinion, quite a bit too many shortcuts have been taken, for example regarding missing i18n, unencrypted storage of private keys and omission of of relevant information in the UI.

There were also several serious security issues:
- generated RSA keys are orders of magnitude weaker than they should be, due to the way they are generated.
- (almost) arbitrary Nimiq messages are blindly signed, which can be used to impersonate the user.
- malicious apps are able to trick the Keyguard into storing and using RSA private keys known to them.

Additionally, this PR points out opportunities for further security hardening, code cleanup, code deduplication, smaller fixes, stricter parsing etc..

danimoh commented:
> @danimoh Do you maybe have a better idea of how to solve the double KeyguardCommand.KeyguardCommand in this commit?
> https://github.com/nimiq/keyguard/commit/253219009034c528395025c8d42168deade52d9c

I tried multiple approaches.
Actually, the nested namespace can be avoided as easily as this:
```typescript
import { KeyguardCommand as _KeyguardCommand } from '../client/src/KeyguardCommand';

export as namespace KeyguardCommand;
export = _KeyguardCommand;
```

But that also occurred to me only later, and I think the approach that I took instead, which is replicating the client enums on the Javascript side of the Keyguard, is eventually nicer. This way, the enums are also usable in the Javascript in a type safe manner, without requiring any type assertions.

I first implemented the replicated enum in https://github.com/nimiq/keyguard/pull/524/commits/1f16c3d10dd181bedc6b164f07e6d747b03352a8, but then simplified / refined the approach in https://github.com/nimiq/keyguard/pull/524/commits/ea4bebed766669a1c949b38c813c7be3e7bbdcff. Both replication solutions are fully type checked and ensure equality with the original enum.

danimoh committed:
Key,Keystore: represent RSA key pair as actual CryptoKeys

danimoh committed:
Key: split and rename functionality of setRsaKeyPair

Improve code readability and use more descriptive function names.

sisou committed:
Rework transaction data formatting

- Allow all transactions to basic accounts to interpret the recipient data as UTF8
- But join it with a contract withdrawal prefix when relevant
- Add a separate translations for adding stake that is not "Staking creation"

sisou committed:
Minor rewordings

sisou committed:
Verify RSA-key-creating forge.min.js file is original

sisou committed:
Move filtering of other signers up

To make it clearer why object-comparison works.

sisou committed:
Re-add guard against same-frame messages

Which are sometimes triggered by browser extensions.

sisou committed:
Remove alias to `SignatureResult` result type

sisou committed:
Add tests for RequestParser.parseLogoUrl's new `allowEmpty` option

sisou committed:
Exclude sandboxed RSA iframe code from test includes

It was throwing an error that it must be run in a sandboxed iframe at the end of the test run.

sisou committed:
Fix and re-enable Key partial signature test

sisou committed:
Fix SignMultisigTransaction demo

sisou committed:
Fix RSAKeys demo

sisou committed:
Fix ConnectAccount demo

sisou committed:
Tabs => 4 spaces in demo files

sisou committed:
Use <script type="module"> instead of using the onload event in demo files

sisou committed:
Clean up TransactionDataFormatting

sisou committed:
Merge pull request #532 from nimiq/soeren/multisig-fixes-fixes

Multisig-fixes Fixes

sisou committed:
Merge pull request #524 from nimiq/daniel/multisig-fixes

Fixes for PR #458 (Multisig: Connect request and SignMultisigTransaction request)
